File: sorting/bubbleSort.java
import java.sql.SQLOutput;
import java.util.Arrays;

public class Main {

    public static void main(String[] args) {
        int[] arr = {5, 10, -3, -10, 1, 100, 99};

        int[] sortedArray = bubbleSort(arr);
        
        //unsorted array remains unsorted because bubbleSort() returns resultArray
        System.out.println(Arrays.toString(arr));

        // Print the sorted array
        System.out.println(Arrays.toString(sortedArray));
    }

    public static int[] bubbleSort(int[] arr) {
        int[] resultArray = Arrays.copyOf(arr, arr.length);

        for (int outerIndex = 0; outerIndex < resultArray.length; outerIndex++) {
            // Using helper variable
            int outerElement = resultArray[outerIndex];

            for (int innerIndex = outerIndex + 1; innerIndex < resultArray.length; innerIndex++) {
                // Using helper variable
                int innerElement = resultArray[innerIndex];

                if (outerElement > innerElement) {
                    resultArray[outerIndex] = innerElement;
                    resultArray[innerIndex] = outerElement;

                    // Update the helper variables
                    outerElement = resultArray[outerIndex];
                    innerElement = resultArray[innerIndex];
                }
            }
        }
        return resultArray;
    }

    //bubble sort without helper variables
    public static int[] bubbleSort2(int[] arr) {
        for (int outerIndex = 0; outerIndex < arr.length; outerIndex++) {
            for (int innerIndex = outerIndex + 1; innerIndex < arr.length; innerIndex++) {
                if (arr[outerIndex] > arr[innerIndex]) {
                    // The 'temp' variable is needed in order to perform the swap
                    int temp = arr[outerIndex];
                    arr[outerIndex] = arr[innerIndex];
                    arr[innerIndex] = temp;
                }
            }
        }
        return arr;
    }
}

File: sorting/bubbleSort.js
function swap(arr, i, j) {
  const temp = arr[i];
  arr[i] = arr[j];
  arr[j] = temp;
}

function bubbleSort(arr) {
  let isSorted;

  for (let i = 0; i < arr.length; i++) {
    isSorted = true;

    for (let j = 0; j < arr.length - 1 - i; j++) {
      if (arr[j] > arr[j + 1]) {
        swap(arr, j, j + 1)
        isSorted = false;
      }
    }

    if (isSorted) {
      break;
    }
  }
  return arr;
}

//Use this version if you want to preserve the original array
//and obtain a sorted version as a separate result.
function bubbleSort2(arr) {
  const resultArr = [...arr];

  for (let i = 0; i < resultArr.length; i++) {
    for (let j = i + 1; j < resultArr.length; j++) {
      if (resultArr[i] > resultArr[j]) {
        // Swap elements
        const temp = resultArr[i];
        resultArr[i] = resultArr[j];
        resultArr[j] = temp;
      }
    }
  }
  return resultArr;
}

const arr = [3, 10, -3, 48, 5, 33, 99];

console.log(bubbleSort2(arr));
